### PR TITLE
Allow `n()` in `data_modify()`

### PR DESCRIPTION
**Pull request by strengejacke**
@etiennebacher Maybe, if checks pass, you can merge this PR for the planned release? Just a minor code change.

### CODE REVIEW

File: NEWS.md
Comment (etiennebacher) [inline NEWS.md:9]:
```suggestion
* `data_modify()` now recognizes `n()`, for example to create an index for data groups 
  with `1:n()` (#535).
```

File: R/data_modify.R
Comment (etiennebacher) [inline R/data_modify.R:25]:
```suggestion
#' - For data frames (including grouped ones), the function `n()` can be used to count the
```

File: R/data_modify.R
Comment (etiennebacher) [inline R/data_modify.R:361]: This feels very hacky and it seems it would break if someone passes `1:fun()` since `n()` will be detected. I also think it would open the door to more "special" functions (`first()`, `last()`, `nth()`, etc.), which can be hard to maintain in the long run.

Random thought: would parsing the expression with getParseData() and using its results make this more robust?

File: tests/testthat/test-data_modify.R
Comment (etiennebacher) [inline tests/testthat/test-data_modify.R:495]: Can you also check how it behaves on 

1. ungrouped dataframe
2. dataframe with 0 rows
3. maybe also inside a function to check that the `eval()` call runs in the right environment?
4. calls that are close but not identical to `n()`, e.g. `fun()`, `1:fun()`, etc.

**Review by etiennebacher — CHANGES_REQUESTED** [review]
Thanks! I think there are cases where this implementation would fail, e.g. if someone passes `1:fun()`. Can you check for this?

You have more usecases than me with `datawizard` but it seems that we gradually try to replace all functions from `dplyr`, which I thought was not the main objective of `datawizard`. I don't know if this needed in another package internally but if not, do we actually need to implement this instead of using `dplyr`?

---

Also, if you want to add more bug fixes/features, I don't plan on releasing before the end of the week.

File: R/data_modify.R
Comment (strengejacke) [reply R/data_modify.R:361]: ok, should no be catching only "n()".

> I also think it would open the door to more "special" functions 

True, but we already have `n()` in `data_summary()`, and I think it's a more common use case than first/last. Thus, this is more a "feature completion" than opening Pandorra's box ;-)

What do you mean by `getParseData()`?

File: tests/testthat/test-data_modify.R
Comment (strengejacke) [reply tests/testthat/test-data_modify.R:495]: All checks added, also behaviour for zero-row data frames.
But I have no good idea for a use-case for `fun()`?

**Comment by strengejacke** [comment]
> but if not, do we actually need to implement this 

You're right, but we had this already for `data_summary()`, because it's an more often used function, and I thought it should then also work for `data_modify()`. That's the main reason why I added this. And actually, I don't use dplyr and tidyr anymore, since 99.5% of my tasks can be done with datawizard. :-)

File: R/data_modify.R
Comment (etiennebacher) [inline R/data_modify.R:155]: Should we error here? I can imagine that this can be used in functions where the input data is not always of the right format and can have 0 rows. I think a better output would be to return a 0-row dataframe with the new columns, like:

``` r
dplyr::mutate(data.frame(x = character(0)), y = 1)
#> [1] x y
#> <0 rows> (or 0-length row.names)
```

My comment was more to ensure that the specific `n()` works as expected when there are 0 rows.

**Review by etiennebacher — CHANGES_REQUESTED** [review]
>  But I have no good idea for a use-case for fun()?

I think you can just check that the error message of `data_modify(mtcars, b = fun())` is comprehensible.

But before that, please check my other comment.

**Comment by etiennebacher** [comment]
> You're right, but we had this already for data_summary(), because it's an more often used function, and I thought it should then also work for data_modify().

I'm also not a fan of having this in `data_summary()`. I think it adds confusion for a `tidyverse` user that could think "why is `n()` supported but not `first()` or `last()`?" and basically we don't have a good answer to that. To me the `tidyverse` and `datawizard` are more complements than substitutes: `datawizard` provides nice statistical functions that can be included in a pipeline, and it also provides dependency-free functions that we need internally in other packages anyway. Adding support for `n()` doesn't seem to meet any of those cases.

File: R/data_modify.R
Comment (strengejacke) [reply R/data_modify.R:155]: Would you prefer just returning an empty data frame then, or give a message/warning and return an empty data frame?

**Comment by strengejacke** [comment]
> > But I have no good idea for a use-case for fun()?
> 
> I think you can just check that the error message of `data_modify(mtcars, b = fun())` is comprehensible.
> 
> But before that, please check my other comment.

But does `fun()` necessarily result in an error? I'm not sure whether there are use cases of user-defined functions?

**Comment by strengejacke** [comment]
> datawizard provides nice statistical functions that can be included in a pipeline, and it also provides dependency-free functions that we need internally in other packages anyway. Adding support for n() doesn't seem to meet any of those cases.

Yes, that was certainly the initial idea of the package, but a) there are probably other people who use datawizard _instead_ of dplyr/tidyr (e.g., @DominiqueMakowski - at least in the past - never really liked the NSE of tidyverse and prefers working with strings to identify variables etc., as do I) and b) maybe there will be other packages that prefer using datawizard for its minimal dependencies in their packages, and then it could be nice to have some features that are also supported by other packages. I personally don't mind implementing "common" features, but what do the other @easystats/maintainers think? Should we have a "stopping rule" of implementing features?

File: R/data_modify.R
Comment (etiennebacher) [reply R/data_modify.R:155]: Actually, the behavior of `base` is to error, while `dplyr` and `data.table` silently return a 0-row dataframe. I think we can follow the base behavior and error here, so forget this comment.

``` r
library(data.table)
library(dplyr, warn.conflicts = FALSE)

df <- data.frame(x = character(0))
dt <- data.table(x = character(0))
tib <- tibble(x = character(0))

df[["y"]] <- sum(1:3)
#> Error in `[[<-.data.frame`(`*tmp*`, "y", value = 6L): replacement has 1 row, data has 0
dt[, y := sum(1:3)][]
#> Empty data.table (0 rows and 2 cols): x,y
tib |> mutate(y = sum(1:3))
#> # A tibble: 0 × 2
#> # ℹ 2 variables: x <chr>, y <int>
```

**Comment by etiennebacher** [comment]
> But does fun() necessarily result in an error? I'm not sure whether there are use cases of user-defined functions?

I just meant that we should have a test for the error below:

``` r
library(datawizard)

head(mtcars) |> 
  data_modify(x = n())
#>                    mpg cyl disp  hp drat    wt  qsec vs am gear carb x
#> Mazda RX4         21.0   6  160 110 3.90 2.620 16.46  0  1    4    4 6
#> Mazda RX4 Wag     21.0   6  160 110 3.90 2.875 17.02  0  1    4    4 6
#> Datsun 710        22.8   4  108  93 3.85 2.320 18.61  1  1    4    1 6
#> Hornet 4 Drive    21.4   6  258 110 3.08 3.215 19.44  1  0    3    1 6
#> Hornet Sportabout 18.7   8  360 175 3.15 3.440 17.02  0  0    3    2 6
#> Valiant           18.1   6  225 105 2.76 3.460 20.22  1  0    3    1 6

head(mtcars) |> 
  data_modify(x = fun())
#> Error: There was an error in the first expression. Attempt to apply
#>   non-function. Possibly misspelled or not yet defined?
```

**Comment by strengejacke** [comment]
@easystats/core-team Any opinions to
https://github.com/easystats/datawizard/pull/535#issuecomment-2316072611
https://github.com/easystats/datawizard/pull/535#issuecomment-2316293369

?

**Comment by rempsyc** [comment]
> I personally don't mind implementing "common" features, but what do the other easystats/maintainers think?

Personally, I think the harm in adding extra features is limited (although I can also see Etienne's point about creating confusion). I think the potential benefit of new features ouweight the cons (although it also adds to future maintenance and development), so as long as you're motivated to do the associated work Daniel (and I think you are), I'd say go ahead 🥸

**Comment by etiennebacher** [comment]
So do we put this in 0.13.0?

**Comment by strengejacke** [comment]
> @easystats/core-team Any opinions to [#535 (comment)](https://github.com/easystats/datawizard/pull/535#issuecomment-2316072611) [#535 (comment)](https://github.com/easystats/datawizard/pull/535#issuecomment-2316293369)
> 
> ?

@easystats/maintainers any further comments on this? :-)

**Comment by rempsyc** [comment]
Add it!

**Review by etiennebacher — APPROVED** [review]
Seems like I'm the only one blocking this but I'm not a big `datawizard` user compared to you so feel free to merge this once the bullet point in NEWS has been put in the good section

**Comment by strengejacke** [comment]
A package maintainer not using the packages (s)he maintains, that's also some kind of punk-rock! 😄 
beavis-and-butthead-headbanging

**Comment by etiennebacher** [comment]
I had some usecases for it when I started contributing but I'm full in tidyverse code for all my projects now and rarely rely on `datawizard` (mostly using the summary / codebook functions) :sweat_smile:

**Pull request merged**
